janitor/db/database.py

```diff
@@ -17,34 +17,40 @@ def __init__(self, creds: DbConnectionDetails):
         Arguments:
             creds {DbConnectionDetails}: details for database connection
         """
-        logger.info(f"Attempting to connect to {creds['host']} on port {creds['port']}...")
+        self.creds = creds
+        self._connection = cast(MySQLConnectionAbstract, None)
+
+    @property
+    def connection(self) -> MySQLConnectionAbstract:
+        """Database connection, attempt to connect if not connected."""
+        if self._connection is not None:
+            return self._connection
+
+        logger.info(f"Attempting to connect to {self.creds['host']} on port {self.creds['port']}...")  # type: ignore
 
         try:
             connection = mysql.connect(
-                host=creds["host"],
-                port=creds["port"],
-                database=creds["db_name"],
-                username=creds["username"],
-                password=creds["password"],
+                host=self.creds["host"],
+                port=self.creds["port"],
+                database=self.creds["db_name"],
+                username=self.creds["username"],
+                password=self.creds["password"],
             )
 
-            if connection is not None:
-                if connection.is_connected():
-                    logger.info(f"MySQL connection to {creds['db_name']} successful!")
-                    self.connection = cast(MySQLConnectionAbstract, connection)
-                    self.cursor = self.connection.cursor()
-                else:
-                    logger.error("MySQL connection failed!")
+            if connection.is_connected():
+                logger.info(f"MySQL connection to {self.creds['db_name']} successful!")
+                self._connection = cast(MySQLConnectionAbstract, connection)
 
         except mysql.Error as e:
             logger.error(f"Exception on connecting to MySQL database: {e}")
 
+        return self._connection
+
     def close(self) -> None:
         """Close connection to database."""
         try:
             if self.connection.is_connected():
```

Comment:
Maybe check that `self._connection is not None and ...` here otherwise you'll make a connection just to close it again if one was never instantiated.